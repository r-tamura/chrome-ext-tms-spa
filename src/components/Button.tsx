import * as React from "react"
import classNames from "~/helpers/classNames"
import { Link } from "react-router-dom"

enum Color {
  DEFAULT          = "",
  PRIMARY          = "primary",
  SECONDARY        = "secondary",
  DANGER           = "danger",
  LINK             = "link",
}

interface IButtonProps extends React.ButtonHTMLAttributes<{}> {
  /* button要素とするかa要素とするか true: button / false: a*/
  button?: boolean
  className?: string
  color?: Color
  /* ボタンをblock要素とするか コンテナコンポーネントに対して最大幅と同じになる */
  block?: boolean
  circle?: boolean
  component?: React.AllHTMLAttributes<{}>
  to?: string
  size?: string
}

const Button: React.SFC<IButtonProps> = ({
  button        = true,
  disabled      = false,
  color         = Color.DEFAULT,
  block         = false,
  circle        = false,
  className: classNameProps,
  component,
  children,
  ...otherProps
} = {}) => {
  const className = classNames(
    button ? "tms-btn" : null,
    color || Color.DEFAULT,
    { disabled, block, circle },
    classNameProps
  )
  const Component = component || button ? "button" : Link

  return (
    <Component className={className} {...otherProps}>
      {children}
    </Component>
  )
}

export {
  IButtonProps,
  Color,
  Button,
}

export default Button
