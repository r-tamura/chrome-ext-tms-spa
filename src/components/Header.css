:root {
  --header-height: 50px;
}

.company-logo {
  display: block;
  width: 40px;
}


.page-header {
  display: flex;
  align-items: center;
  flex-direction: row;
  padding: 0 5px;
  background-color: var(--primary-color-reverse);
  /* color: var(--white); */
  font-size: var(--font-size-lg);
  font-weight: bolder;
  line-height: var(--header-height);
  
  @apply --z-index-high;
  @apply --box-shadow;

  & a:focus, & a:hover {
    opacity: 0.8
  }

  & .nav {
    height: 80%;
    margin: 5px 0;
  }

  & .nav-link {
    height: 100%;

    &.active {
      background-color: $nav-selected-bg-color;
    }
  }

  @media (--md) {
    /* ロゴ非表示 */
    .nav, .nav-link {
        display: none;
    }

    .menu-btn {
        display: block;
    }
  }
}

/* ヘッダーメニューボタン */
.menu-btn {
    cursor: pointer;
    padding: $header-height/4 10px;
    display: none;
    &:hover {
      background-color: darken($COLOR_HEADER_BG, 10%);
      border-radius: 50%;
    }

    img {
      height: 50%;
    }
}

.header-user {
  /* @extend .clickable; */
  display: flex;
  align-items: center;
  margin-left: auto;

  .person-icon {
    background-image: url(chrome-extension://odbcdehjjdjjnpenjbblpmkkibggblpe/chrome/assets/images/person.svg);
    height: 50%;
    width: 25px;
    display: block;
  }

  &:hover {
    text-decoration: underline;
  }
}

.header-user, .nav-link {
  img {
    display: inline-block;
    height: $header-height;
    width: $header-height;
    padding: $header-height * 0.10 10px;
  }
}

.header-user .dropdown-menu {
  right: 0px;
}
