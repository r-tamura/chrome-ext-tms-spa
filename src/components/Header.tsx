import * as React from "react"
import { connect } from "react-redux"
import { RootState } from "~/modules"

interface IProps extends React.ClassAttributes<{}> {
  username: string
  // isLogin: boolean
}

  // function getInitState(props) {
  //   return {
  //     isLogin: ContextStore.getIsLogin(),
  //     username: ContextStore.getName(),
  //   }
  // }

  /* メニューオープン3点リーダクリックイベント */
const onMenuBtnClick = (e: React.MouseEvent<HTMLElement>) => {
  const nav = document.getElementById("side-nav-bar")
  nav.classList.toggle("on-active")
}

// function onSelectTabClick(e: React.SyntheticEvent) {
//   // "+"でstring=>number変換
//   const id = +(e.currentTarget as HTMLElement).getAttribute("data-active-id")
//   CommonAPI.updateMainTab(id)
// }

// /* ユーザボタンクリックイベント */
// function onUserClick = (e: React.SyntheticEvent) => {
//   /* ユーザメニューを開く */
//   (refs["usermenu"] as UserMenu).show()
// }

// updateLoginState() {
//   let isLogin = ContextStore.getIsLogin()
//   let username = ContextStore.getName()
//   setState({isLogin, username})
// }

// ユーザJSXを生成します
function renderUser(username: string) {
  let loginButton
  const isLogin = username && username.length > 0
  if (isLogin) {
    loginButton = (
      <div className="header-user" onClick={() => ({})}>
        <i className="person-icon"/>
        <span>{username}</span>
        {/* <UserMenu ref="usermenu"/> */}
      </div>
    )
  }
  return loginButton
}

// リンク一覧を生成します
function renderNavLinks(): JSX.Element[] {

  // ログイン前はリンクを表示しない
  return [<div key={1} className="displayNone"/>]

  // const links = [
  //   { to: "transportation", image: taxiSvg, title: "交通費"},
  //   { to: "attendance", image: bookSvg, title: "月間勤怠"},
  // ]
  // return (
  //   links
  //     .map((e, i) => {
  //       const classActive = e.to === props.activeId ? "active" : ""
  //       return (
  //         <div key={e.to} className={`nav-link ${classActive}`}>
  //         <Link to={e.to} title={e.title}>
  //           {e.image}
  //         </Link>
  //         </div>
  //       )
  //     }
  //   )
  // )
}

/**
 * アプリケーションヘッダーコンポーネント
 */
const Header: React.SFC<IProps> = ({
  username,
}) => {

    // const { isLogin } = props
    const isLogin = true
    const navLinks = renderNavLinks()
    const user = renderUser(username)

    return (
      <header className="page-header">
        {/* メニューボタン 小サイズデバイスで表示 */}
        { isLogin &&
          <div className="menu-btn" onClick={onMenuBtnClick}>
            <img src="" />
          </div>
        }
        <a className="company-logo" href="http://www.telema.jp/">
          <img src="https://www.telema.co.jp/images/logo_sp.svg"/>
        </a>
        {/* リンク */}
        {navLinks}

        {/* ユーザ */}
        {user}
      </header>
    )
  }

type OwnProps = Partial<IProps>

const mapStateToProps = (state: RootState, ownProps: OwnProps) => {
  return {
    username: state.user.name,
  }
}

export default connect(mapStateToProps)(Header)
