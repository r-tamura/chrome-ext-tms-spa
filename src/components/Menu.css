.menu-main {
  position: absolute;
  min-width: 16px;
  min-height: 16px;
  overflow-y: auto;
  overflow-x: hidden;
  border-radius: 2px;
  background-color: #fff;
  opacity: 1;
  transform: scale(1, 1) translate(-50px, 57px);
  /* transform-origin: 0px 32px; */
  transition:
    opacity 267ms 
    cubic-bezier(0.4, 0, 0.2, 1) 0ms,
    transform 178ms 
    cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  box-shadow:
    0px 5px 5px -3px rgba(0, 0, 0, 0.2),
    0px 8px 10px 1px rgba(0, 0, 0, 0.14),
    0px 3px 14px 2px rgba(0, 0, 0, 0.12);
}

.menu-list {
  padding-top: 8px;
  padding-bottom: 8px;
}

.menu-item {
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 12px;
  padding-bottom: 12px;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5em;
  white-space: nowrap;
  text-overflow: ellipsis;
  cursor: pointer;

  &:hover {
    background-color: rgba(0, 0, 0, 0.14);
  }

  &.menu-item--strong {
    cursor: text;
    font-weight: bolder;

    &:hover {
      background-color: initial;
    }
  }
}


