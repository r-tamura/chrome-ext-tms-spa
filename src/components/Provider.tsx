import React from "react";
import { ThemeProvider } from "styled-components";
import { telema } from "~/styles/theme";

export const AppThemeProvider: React.SFC<{ children: JSX.Element }> = ({
  children
}) => <ThemeProvider theme={telema}>{children}</ThemeProvider>;
