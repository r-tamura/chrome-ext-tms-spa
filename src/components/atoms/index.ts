export * from "./Button";
export * from "./Panel";
export * from "./Modal";
export * from "./Form";
export * from "./TextField";
