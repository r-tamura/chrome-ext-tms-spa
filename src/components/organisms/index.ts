export * from "./AppGrid";
export * from "./MainContent";
export * from "./Header";
export * from "./TransExpenseHistory";
export * from "./TransExpenseTemplates";
export * from "./Nav";
