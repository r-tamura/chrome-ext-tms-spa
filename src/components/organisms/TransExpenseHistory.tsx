import React, { useState } from "react";
import { find, propEq } from "ramda";
import {
  Button,
  ButtonGroup,
  Panel,
  RequiredInput,
  Modal
} from "~/components/atoms";
import { AddIcon } from "~/components/Icon";
import { CURRENCY } from "~/helpers/_const";
import {
  Project,
  Usage,
  Objective,
  TransExpenseView,
  TransExpenseTemplateView,
  ExpenseId,
  TransExpenseCreateRequest,
  TransExpenseUpdateRequest,
  TransExpense
} from "~/types";

import { Table, Tr, Td } from "~/components/atoms/Table";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faEdit, faTrash } from "@fortawesome/free-solid-svg-icons";
import { useModal } from "~/components/hooks";
import useForm from "react-hook-form";
import { SelectBox } from "../SelectBox";

interface IProps {
  expenses: TransExpenseView[];
  templates: TransExpenseTemplateView[];
  projects: Project[];
  usages: Usage[];
  objectives: Objective[];
  deleteExpense: (expenseId: ExpenseId) => void;
  createExpense: (expense: TransExpenseCreateRequest) => void;
  updateExpense: (expense: TransExpenseUpdateRequest) => void;
  createExpenseFromTemplate: (templateId: string, strdate: string) => void;
}

interface IState {
  selected?: ExpenseId; // 選択中の交通費ID
  useTemplate?: boolean;
}

/**
 * 登録済み交通費リストセクション
 */
export const TransExpenseHistory: React.FC<IProps> = props => {
  const { isOpen: isModalOpen, open, close } = useModal(false);
  const [state, setState] = useState(getInitState(props));

  const { expenses, deleteExpense } = props;
  const { useTemplate } = state;
  return (
    <div>
      {expenses.length > 0 ? <Expenses /> : <p>交通費が登録されていません。</p>}
      <ButtonGroup>
        <Button
          variant={"contained"}
          color={"secondary"}
          onClick={openCreateFromTemplateModal}
        >
          テンプレートから新規登録
        </Button>
        <Button
          id={"create-trans-expense"}
          variant={"circle"}
          color={"primary"}
          onClick={openCreateModal}
        >
          <AddIcon />
        </Button>
      </ButtonGroup>

      {/* 編集モーダル */}
      <EditorModal />
    </div>
  );

  function Expenses() {
    const header = [
      "#",
      "利用日",
      "プロジェクト",
      "利用区分",
      "目的",
      "金額",
      "",
      ""
    ];
    return (
      <div>
        <Table header={header}>
          {expenses.map((e, i) => (
            <Tr key={(() => e.expenseId)()} hover>
              <Td className="record-id">{i + 1}</Td>
              <Td>{e.strdate}</Td>
              <Td>{e.project.name}</Td>
              <Td>{e.usage.name}</Td>
              <Td>{e.objective.name}</Td>
              <Td>
                <strong>{CURRENCY.format(e.cost)}</strong>
              </Td>
              <Td>
                <Button
                  variant="icon"
                  color="primary"
                  onClick={() => openEditModal(e.expenseId)}
                  title="編集"
                >
                  <FontAwesomeIcon icon={faEdit} />
                </Button>
              </Td>
              <Td>
                <Button
                  variant="icon"
                  color="danger"
                  onClick={() => deleteExpense(e.expenseId.toString())}
                  title="削除"
                >
                  <FontAwesomeIcon icon={faTrash} />
                </Button>
              </Td>
            </Tr>
          ))}
        </Table>
      </div>
    );
  }

  /*
   * 交通費新規作成/編集モーダルをレンダーします
   */
  function EditorModal() {
    type FormData = Record<keyof TransExpense, string>;
    const {
      register,
      handleSubmit,
      clearError,
      errors,
      setError,
      formState
    } = useForm<FormData>();
    const {
      expenses,
      projects,
      usages,
      objectives,
      createExpense,
      updateExpense
    } = props;
    const { selected } = state;

    function handleCancelClick(e: React.MouseEvent) {
      e.preventDefault();
      e.stopPropagation();
      closeModal();
    }

    function onSubmit({
      strdate,
      projectId,
      usageId,
      objectiveId,
      customer,
      from,
      to,
      cost: costStr
    }: FormData) {
      let cost;
      cost = Number.parseInt(costStr, 10);
      if (Number.isNaN(cost)) {
        setError(
          "cost",
          "NaN",
          "有効な金額ではありません、正の整数値を入力してください"
        );
        return;
      }
      if (cost < 0) {
        setError("cost", "NaN", "正の整数値を入力してください");
        return;
      }

      if (selected) {
        updateExpense({
          expenseId,
          strdate,
          projectId,
          usageId,
          objectiveId,
          customer,
          from,
          to,
          cost
        });
      } else {
        createExpense({
          strdate,
          projectId,
          usageId,
          objectiveId,
          customer,
          from,
          to,
          cost
        });
      }
      closeModal();
    }

    const defaultExpense: TransExpenseView = {
      expenseId: null,
      strdate: "",
      project: projects[0],
      usage: usages[0],
      objective: objectives[0],
      customer: "",
      from: "",
      to: "",
      cost: 0
    };

    const {
      expenseId,
      strdate,
      project,
      usage,
      objective,
      customer,
      from,
      to,
      cost
    }: TransExpenseView =
      find(propEq("expenseId", selected), expenses) || defaultExpense;
    const costStr = cost.toString();
    return (
      <Modal isOpen={isModalOpen} onRequestClose={() => close()}>
        <Panel>
          <form onSubmit={handleSubmit(onSubmit)}>
            <RequiredInput
              name={"strdate"}
              label={"日付"}
              placeholder={"ex: 20190401"}
              error={errors.strdate}
              defaultValue={strdate}
              register={register}
              clearError={clearError}
            />
            <SelectBox
              name="projectId"
              label="プロジェクトコード *"
              defaultValue={project.projectId}
              options={{
                items: projects,
                value: (project: Project) => project.projectId,
                label: (project: Project) => project.name
              }}
              ref={register({ required: true })}
            />
            <SelectBox
              name="usageId"
              label="利用区分 *"
              defaultValue={usage.usageId}
              options={{
                items: usages,
                value: (usage: Usage) => usage.usageId,
                label: (usage: Usage) => usage.name
              }}
              ref={register({ required: true })}
            />
            <SelectBox
              name="objectiveId"
              label="目的コード *"
              defaultValue={objective.objectiveId}
              options={{
                items: objectives,
                value: (objective: Objective) => objective.objectiveId,
                label: (objective: Objective) => objective.name
              }}
              ref={register({ required: true })}
            />
            <RequiredInput
              name={"customer"}
              label={"顧客先"}
              placeholder={"ex: XXXX 株式会社"}
              error={errors.customer}
              defaultValue={customer}
              register={register}
              clearError={clearError}
            />
            <RequiredInput
              name={"from"}
              label={"乗車地"}
              placeholder={"ex: JR 千葉駅"}
              error={errors.from}
              defaultValue={from}
              register={register}
              clearError={clearError}
            />
            <RequiredInput
              name={"to"}
              label={"降車地"}
              placeholder={"ex: JR 東京駅"}
              error={errors.to}
              defaultValue={to}
              register={register}
              clearError={clearError}
            />
            <RequiredInput
              name={"cost"}
              label={"金額"}
              placeholder={"ex: 500"}
              error={errors.cost}
              defaultValue={costStr}
              register={register}
              clearError={clearError}
            />
            <ButtonGroup>
              <Button variant="contained" onClick={handleCancelClick}>
                キャンセル
              </Button>
              <Button
                variant="contained"
                color="primary"
                disabled={!formState.isValid}
              >
                {expenseId ? "更新" : "作成"}
              </Button>
            </ButtonGroup>
          </form>
        </Panel>
      </Modal>
    );
  }

  function renderCreateFromTemplateModal() {
    const { templates, createExpenseFromTemplate } = props;
    return (
      // <Modal
      //   isOpen={isModalOpen}
      //   formItems={[
      //     {
      //       type: FormItemType.TEXT,
      //       name: "strdate",
      //       label: "日付",
      //       value: ""
      //     },
      //     {
      //       type: FormItemType.SELECT,
      //       name: "templateId",
      //       label: "テンプレート名",
      //       value: "",
      //       options: {
      //         items: templates,
      //         valueKey: "templateId",
      //         labelKey: "templateName"
      //       }
      //     }
      //   ]}
      //   onClose={closeModal}
      //   onOKClick={({
      //     templateId,
      //     strdate
      //   }: {
      //     templateId: string;
      //     strdate: string;
      //   }) => createExpenseFromTemplate(templateId, strdate)}
      // />
      <div> history template</div>
    );
  }

  function getInitState(props: IProps): IState {
    return { useTemplate: false, selected: undefined };
  }

  /* モーダルクローズ */
  function closeModal() {
    setState({
      useTemplate: false,
      selected: null
    });
    close();
  }

  /* 編集モーダル */
  function openEditModal(expenseId: ExpenseId) {
    setState({
      useTemplate: false,
      selected: expenseId
    });
    open();
  }

  /* 新規登録モーダル */
  function openCreateModal() {
    setState({
      useTemplate: false,
      selected: null
    });
    open();
  }

  /* テンプレートから新規登録モーダル */
  function openCreateFromTemplateModal() {
    setState({
      useTemplate: true,
      selected: null
    });
    open();
  }
};
