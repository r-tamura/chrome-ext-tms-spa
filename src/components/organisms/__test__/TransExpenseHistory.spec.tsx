import React from "react";
import { shallow, mount, ReactWrapper } from "enzyme";
import { TransExpenseHistory } from "../TransExpenseHistory";
import toJson from "enzyme-to-json";
import { AppThemeProvider } from "~/components/Provider";
import { TransExpenseView } from "~/types";
import { setValue, wait } from "~/components/test_util";

function clickCreateButton(wrapper: ReactWrapper) {
  wrapper
    .find("button")
    .last()
    .simulate("click");
}

function clickCreateFromTemplateButton(wrapper: ReactWrapper) {
  wrapper.find(`button[color="secondary"]`).simulate("click");
}

describe("<TransExpenseHistory />", () => {
  it("shallow render snapshot", () => {
    const spyCreate = jest.fn();
    const spyCreateFromTemplate = jest.fn();
    const spyUpdate = jest.fn();
    const spyDelete = jest.fn();
    expect(
      toJson(
        shallow(
          <TransExpenseHistory
            expenses={[]}
            templates={[]}
            projects={[]}
            usages={[]}
            objectives={[]}
            createExpense={spyCreate}
            createExpenseFromTemplate={spyCreateFromTemplate}
            updateExpense={spyUpdate}
            deleteExpense={spyDelete}
          />
        )
      )
    ).toMatchSnapshot();
  });

  describe("mount with no expenses", () => {
    const spyCreate = jest.fn();
    const spyCreateFromTemplate = jest.fn();
    const spyUpdate = jest.fn();
    const spyDelete = jest.fn();
    const $ = mount(
      <AppThemeProvider>
        <TransExpenseHistory
          expenses={[]}
          templates={[
            {
              templateId: "xxxx-yyyy",
              templateName: "Template A",
              customer: "XXXX K.K.",
              project: { name: "Project X", projectId: "201108010001" },
              usage: { name: "交通費(往復)", usageId: "3" },
              objective: { name: "作業", objectiveId: "2" },
              from: "Stop A",
              to: "Stop B",
              cost: 1000
            }
          ]}
          projects={[{ name: "Project A", projectId: "1234" }]}
          usages={[{ name: "交通費(片道)", usageId: "1" }]}
          objectives={[{ name: "作業", objectiveId: "1" }]}
          createExpense={spyCreate}
          createExpenseFromTemplate={spyCreateFromTemplate}
          updateExpense={spyUpdate}
          deleteExpense={spyDelete}
        />
      </AppThemeProvider>
    );

    it("登録されていない旨のメッセージ表示する", () => {
      expect($.find("p").text()).toBe("交通費が登録されていません。");
    });

    it("新規作成ボタンを押したときに、入力フォームを表示する", () => {
      expect($.find("form").length).toBe(0);
      $.find("button#create-trans-expense").simulate("click");
      expect($.find("form").length).toBe(1);
    });
  });

  describe("mount with some expenses", () => {
    const spyCreate = jest.fn();
    const spyCreateFromTemplate = jest.fn();
    const spyUpdate = jest.fn();
    const spyDelete = jest.fn();
    const fakeExpenses: TransExpenseView[] = [
      {
        expenseId: "7980",
        strdate: "20190701",
        customer: "XXXX株式会社",
        from: "JR 東京駅",
        to: "JR 新宿駅",
        cost: 777,
        project: {
          name: "Project X",
          projectId: "201108010001"
        },
        usage: {
          name: "交通費(往復)",
          usageId: "3"
        },
        objective: {
          name: "作業",
          objectiveId: "2"
        }
      },
      {
        expenseId: "7981",
        strdate: "20190720",
        customer: "YYYY株式会社",
        from: "JR 津田沼駅",
        to: "JR 千葉駅",
        cost: 211,
        project: {
          name: "Project Y",
          projectId: "201108010002"
        },
        usage: {
          name: "交通費(片道)",
          usageId: "1"
        },
        objective: {
          name: "打ち合わせ",
          objectiveId: "1"
        }
      }
    ];
    const $ = mount(
      <AppThemeProvider>
        <TransExpenseHistory
          expenses={fakeExpenses}
          templates={[
            {
              templateId: "xxxx-yyyy",
              templateName: "Template A",
              customer: "XXXX K.K.",
              project: { name: "Project X", projectId: "201108010001" },
              usage: { name: "交通費(往復)", usageId: "3" },
              objective: { name: "作業", objectiveId: "2" },
              from: "Stop A",
              to: "Stop B",
              cost: 1000
            }
          ]}
          projects={[{ name: "Project X", projectId: "201108010001" }]}
          usages={[{ name: "交通費(往復)", usageId: "3" }]}
          objectives={[{ name: "作業", objectiveId: "2" }]}
          createExpense={spyCreate}
          createExpenseFromTemplate={spyCreateFromTemplate}
          updateExpense={spyUpdate}
          deleteExpense={spyDelete}
        />
      </AppThemeProvider>
    );

    it("登録された交通費データ分のデータを表示する", () => {
      expect($.find("tbody tr").length).toBe(2);
    });

    it("デフォルトでは編集モーダルは表示しない", () => {
      expect($.find("form").length).toBe(0);
    });

    describe("作成モーダル", () => {
      beforeEach(() => {
        clickCreateButton($);
      });

      describe("表示", () => {
        it("編集フォーム", () => {
          expect($.find("form").length).toBe(1);
        });

        it("作成ボタン", () => {
          const $buttons = $.find("form button");
          expect($buttons.last().text()).toBe("作成");
        });
      });

      describe("未入力項目なしでサブミット", () => {
        beforeAll(() => {
          clickCreateButton($);

          setValue($, "strdate", "20191231");
          setValue($, "customer", "サンプル株式会社");
          setValue($, "from", "JR錦糸町");
          setValue($, "to", "JR東京駅");
          setValue($, "cost", "800");
          $.find("form").simulate("submit");
        });

        it("交通費作成関数が呼び出される", async () => {
          // TODO: useFormのhandleSubmitが非同期? 一定時間待つ必要がある(要調査)
          await wait();
          expect(spyCreate.mock.calls.length).toBe(1);
          expect(spyCreate.mock.calls[0][0]).toEqual({
            strdate: "20191231",
            customer: "サンプル株式会社",
            from: "JR錦糸町",
            to: "JR東京駅",
            cost: 800,
            projectId: "201108010001",
            usageId: "3",
            objectiveId: "2"
          });
        });
      });
    });

    describe("編集モーダル", () => {
      beforeAll(() => {
        $.find("tbody tr")
          .first()
          .find("button")
          .first()
          .simulate("click");
      });

      describe("表示", () => {
        it("編集フォーム", () => {
          expect($.find("form").length).toBe(1);
        });

        it("日付", () => {
          const $input = $.find(`input[name="strdate"]`);
          expect($input.length).toBe(1);
          expect($input.prop("defaultValue")).toBe("20190701");
        });

        it("プロジェクト", () => {
          expect($.find(`select[name="projectId"]`).length).toBe(1);
        });

        it("利用区分", () => {
          expect($.find(`select[name="usageId"]`).length).toBe(1);
        });

        it("目的区分", () => {
          expect($.find(`select[name="objectiveId"]`).length).toBe(1);
        });
        it("顧客先", () => {
          const $input = $.find(`input[name="customer"]`);
          expect($input.length).toBe(1);
          expect($input.prop("defaultValue")).toBe("XXXX株式会社");
        });
        it("乗車地", () => {
          const $input = $.find(`input[name="from"]`);
          expect($input.length).toBe(1);
          expect($input.prop("defaultValue")).toBe("JR 東京駅");
        });
        it("降車地", () => {
          const $input = $.find(`input[name="to"]`);
          expect($input.length).toBe(1);
          expect($input.prop("defaultValue")).toBe("JR 新宿駅");
        });
        it("コスト", () => {
          const $input = $.find(`input[name="cost"]`);
          expect($input.length).toBe(1);
          expect($input.prop("defaultValue")).toBe("777");
        });

        it("更新ボタン", () => {
          const $buttons = $.find("form button");
          expect($buttons.last().text()).toBe("更新");
        });

        it("モーダルを閉じる", () => {
          const $buttons = $.find("form button");
          $buttons.first().simulate("click");
          expect($.find("form").length).toBe(0);
        });
      });

      describe("未入力項目あり", () => {
        beforeAll(() => {
          $.find("tbody tr")
            .first()
            .find("button")
            .first()
            .simulate("click");
          $.find("form button")
            .last()
            .simulate("click");
        });

        it("更新サブミットされない", () => {
          expect(spyUpdate.mock.calls.length).toBe(0);
        });
      });

      describe("未入力項目なし", () => {
        beforeAll(() => {
          $.find("tbody tr")
            .first()
            .find("button")
            .first()
            .simulate("click");

          setValue($, "strdate", "20190901");
          setValue($, "customer", "サンプル株式会社");
          setValue($, "from", "JR 新宿駅");
          setValue($, "to", "JR 横浜駅");
          setValue($, "cost", "500");
          $.find("form").simulate("submit");
        });

        it("編集フォーム", () => {
          expect($.find("form").length).toBe(1);
        });

        it("更新関数が呼び出される", async () => {
          // TODO: useFormのhandleSubmitが非同期? 一定時間待つ必要がある(要調査)
          await wait();
          expect(spyUpdate.mock.calls.length).toBe(1);
          expect(spyUpdate.mock.calls[0][0]).toEqual({
            expenseId: "7980",
            strdate: "20190901",
            customer: "サンプル株式会社",
            from: "JR 新宿駅",
            to: "JR 横浜駅",
            cost: 500,
            projectId: "201108010001",
            usageId: "3",
            objectiveId: "2"
          });
        });
      });
    });

    describe("削除", () => {
      it("削除関数が呼び出される", () => {
        $.find("tbody tr")
          .first()
          .find("button")
          .last()
          .simulate("click");
        expect(spyDelete.mock.calls.length).toBe(1);
        expect(spyDelete.mock.calls[0][0]).toBe("7980");
      });
    });

    describe("作成モーダル(交通費テンプレート)", () => {
      beforeAll(() => {
        clickCreateFromTemplateButton($);
      });

      it("フォーム", () => {
        expect($.find("form").length).toBe(1);
      });

      it("日付", () => {
        const $input = $.find(`input[name="strdate"]`);
        expect($input.length).toBe(1);
        expect($input.prop("defaultValue")).toBe("");
      });

      it("テンプレート", () => {
        const $select = $.find(`select[name="templateId"]`);
        expect($select.length).toBe(1);
        expect($select.prop("defaultValue")).toBe("xxxx-yyyy");
      });

      describe("未入力項目なしでフォームサブミット", () => {
        beforeAll(() => {
          setValue($, "strdate", "20191001");
          $.find("form").simulate("submit");
        });

        it("交通費作成関数(テンプレート)が呼び出される", async () => {
          await wait();
          expect(spyCreateFromTemplate.mock.calls.length).toBe(1);
          expect(spyCreateFromTemplate.mock.calls[0]).toEqual([
            "xxxx-yyyy",
            "20191001"
          ]);
        });

        // Note: enzyme上だとform要素が残る 要調査
        // it("モーダルが閉じられる", async () => {
        //   await wait();
        //   expect($.find("form").length).toBe(0);
        // });
      });
    });
  });
});
