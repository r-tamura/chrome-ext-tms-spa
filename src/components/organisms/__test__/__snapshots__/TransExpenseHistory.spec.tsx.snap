// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TransExpenseHistory /> shallow render snapshot 1`] = `
<div>
  <p>
    交通費が登録されていません。
  </p>
  <styled.div>
    <Component
      color="secondary"
      onClick={[Function]}
      variant="contained"
    >
      テンプレートから新規登録
    </Component>
    <Component
      color="primary"
      id="create-trans-expense"
      onClick={[Function]}
      variant="circle"
    >
      <Component />
    </Component>
  </styled.div>
  <EditorModal />
  <CreateFromTemplateModal />
</div>
`;
